File: DSA /Unit 05/Lec 34/Q3.cpp
Single File Programming Question
Problem Statement



Arjun is a database administrator who needs to store a set of unique ID numbers into a fixed-size hash table for quick retrieval. When a collision occurs (two IDs hash to the same index), he must use quadratic probing to find the next available slot. Help him to implement the task.



FORMULA

Hash Value Calculation: hv = arr[i] % tsize
Quadratic Probing (on collision): new_index = (hv + j * j) % tsize where j = 1, 2, 3, ... until an empty slot is found.
Input format :
The first line of input consists of an integer n, representing the number of elements to be inserted.

The second line of input consists of n space-separated integers arr, representing the elements to be inserted into the hash table.

The third line of input consists of an integer tsize, representing the size of the hash table.

Output format :
The first line of output prints: "Hash table after quadratic probing: " followed by tsize space-separated integers representing the final state of the hash table (with -1 indicating empty slots).



Refer to the sample output for formatting specifications.

Code constraints :
The given test cases fall under the following constraints:

1 ≤ n ≤ 100

0 ≤ arr[i] ≤ 10000

1 ≤ tsize ≤ 100

Sample test cases :
Input 1 :
5
12 15 20 25 30
7
Output 1 :
Hash table after quadratic probing: -1 15 30 -1 25 12 20 
Input 2 :
3
13 16 18
5
Output 2 :
Hash table after quadratic probing: -1 16 -1 13 18



#include <bits/stdc++.h>
using namespace std;

int main() {
    int n;
    cin >> n;

    vector<int> arr(n);
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }

    int tsize;
    cin >> tsize;

    vector<int> table(tsize, -1);

    for (int i = 0; i < n; i++) {
        int hv = arr[i] % tsize;

        if (table[hv] == -1) {
            table[hv] = arr[i];
        } else {
            int j = 1;
            while (j < tsize) {
                int new_index = (hv + j * j) % tsize;
                if (table[new_index] == -1) {
                    table[new_index] = arr[i];
                    break;
                }
                j++;
            }
        }
    }

    cout << "Hash table after quadratic probing: ";
    for (int i = 0; i < tsize; i++) {
        cout << table[i] << " ";
    }

    return 0;
}

File: Java/Unit 03/Static/Q3.java
Single File Programming Question
Problem Statement





Akash has been assigned the task of developing a program to determine whether three given lengths can form a valid triangle. A triangle is considered valid if the sum of any two sides is greater than the third side. To implement this, Akash needs to create a program that takes three integer inputs representing the side lengths and checks whether they satisfy the triangle inequality conditions: a + b > c, a + c > b, and b + c > a. 



The solution should be designed using inheritance, incorporating features such as overriding the toString() and equals() methods, utilizing the instanceof operator, or using super and final keywords.

Input format :
The input consists of three space-separated integers, each representing the length of a side of a triangle.

Output format :
The output displays a single line indicating whether the given side lengths form a triangle or not.

If the side lengths form a triangle, it prints "Triangle".

If the side lengths do not form a triangle, it prints "Not a Triangle".



Refer to the sample output for the formatting specifications.

Code constraints :
The given testcases fall under the following constraints:

1 ≤ side 1, side 2, side 3 ≤ 60

Sample test cases :
Input 1 :
3 4 5
Output 1 :
Triangle
Input 2 :
5 12 30
Output 2 :
Not a Triangle
Input 3 :
35 35 60
Output 3 :
Triangle
Whitelist
Set 1:
toString
equals


import java.util.Scanner;

class Shape {
    int a, b, c;

    Shape(int a, int b, int c) {
        this.a = a;
        this.b = b;
        this.c = c;
    }
}

class Triangle extends Shape {

    Triangle(int a, int b, int c) {
        super(a, b, c);
    }

    void checkTriangle() {
        if ((a + b > c) && (a + c > b) && (b + c > a)) {
            System.out.println("Triangle");
        } else {
            System.out.println("Not a Triangle");
        }
    }

    public String toString() {
        return a + " " + b + " " + c;
    }

    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (!(obj instanceof Triangle)) return false;

        Triangle t = (Triangle) obj;
        return a == t.a && b == t.b && c == t.c;
    }
}
public class Main {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        int side1 = input.nextInt();
        int side2 = input.nextInt();
        int side3 = input.nextInt();

        Triangle t = new Triangle(side1, side2, side3);
        t.checkTriangle();

        input.close();
    }
}
